File: python/Memento/Originator.py
from IMemento import IMemento
import Memento
from IOriginator import IOriginator
import string

class Originator(IOriginator):

    def __init__(self, state: int) -> None:
        print("ORIGINATOR: Creating originator")
        self.state = state
        self.name = "power x 1"
        self.counter = 1            

    #creating memento initialized with originator internal state
    def CreateBackup(self) -> IMemento:
        print("ORIGINATOR: Saving state")
        return Memento.Memento(self, self.state, self.name)
    
    #setting originator internal state
    def SetState(self, state: int, name: string) -> None:
        self.state = state
        self.name = name
        self.counter -= 1

    #simulating important logic operations
    def DoubleState(self) -> None:
        self.state *= 2
        self.counter += 1
        self.name = f"power x {str(self.counter)}"

    def ShowState(self) -> string:
        return f"State: {str(self.state)} Name: {self.name}"

File: C#/Memento/Originator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Memento.Memento
{
    class Originator : IOriginator
    {
        private int state;
        private string name;
        private int counter;
        private DateTime date;

        public Originator(int state)
        {
            Console.WriteLine("ORIGINATOR: Creating originator");
            this.state = state;
            name = "multiplier x 1";
            counter = 1;
            date = DateTime.Now;
        }

        //creating memento initialized with originator internal state
        public IMemento CreateBackup()
        {
            Console.WriteLine("ORIGINATOR: Saving state");
            return new Memento(this, state, name, date);

        }

        //setting originator internal state
        public void SetState(int state, string name)
        {
            this.state = state;
            this.name = name;
        }        

        //simulating important logic operations
        public void DoubleState()
        {
            //Console.WriteLine("ORIGINATOR: Changing state value");
            state *= 2;
            counter++;
            name = "multiplier x " + counter.ToString();
        }

        public override string ToString()
        {
            return "State: " + state.ToString() + " Name: " + name.ToString();
        }
    }
}
